### PR TITLE
[Participant] Pipelined transaction indexing

### PR DESCRIPTION
**Pull request by tudor-da**
Currently, the participant indexer processes `TransactionAccepted` atomically. More specifically, all the SQL updates to the affected tables are done in one transaction (`participant_events`, `participant_contracts`, `participant_contract_witnesses`, `participant_command_completions` and `parameters` - for ledger end updates). 

With the purpose of increasing the indexer throughput (by higher parallelism to the index DB), this PR breaks the aforementioned DAML transaction into multiple SQL transactions, updating the involved tables in separate async Akka operators:
* events update (`participant_events`)
* state update (`participant_contracts` and `participant_contract_witnesses`)
* transaction completion (`participant_command_completions` and `parameters` - ledger end update)

### Involved changes
* There will be two supported indexer behaviors: `AtomicExecuteUpdate` (legacy behavior - used for DAML-on-SQL and H2 `JdbcIndexer`) and `PipelinedExecuteUpdate` which implements the mentioned optimization.
* `JdbcIndexer` becomes agnostic of the `JdbcLedgerDao` - it delegates the dao updates to the `ExecuteUpdate` flows. 
* In order to support mid-update crashes (and automatic restarts), `participant_events` insertions become idempotent.
* For consistency reasons, `get(Flat)TransactionBy(Event)Id` queries become guarded by the current ledger end.
* Historically, the ledger entry insertion performed atomically was timed with `daml_index_db_store_ledger_entry`. When using the pipelined insertion, the single metric measures  all three stages combined (across Akka stage boundaries). Additionally, each stage is timed independently by `daml_index_db_store_ledger_entry_events`, `daml_index_db_store_ledger_entry_state` and `daml_index_db_store_ledger_entry_completion`.

### Performance improvement
The benchmarked performance improvement is ~20 %, but expected to be higher with other optimizations affecting events insertion latency.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Review by SamirTalwar-DA — APPROVED** [review]
This looks good to me. Please get a review from someone more familiar with the code before merging.

Follow-up comment coming soon with some ideas for how to improve the safety of this going forward.

**Comment by SamirTalwar-DA** [comment]
The tests timed out on macOS.

**Comment by SamirTalwar-DA** [comment]
I worry that we will accidentally introduce writes to the same table across transactions, or as you pointed out, reads that join tables that span two of the transactions.

Perhaps we could make this safer by creating aggregates that gate access to collections of tables that should always be internally consistent. For example, a "completions" aggregate could manage access to the completions and parameters tables. These aggregates would be responsible for creating the database transaction and passing in the references to the tables, which would mean it would be impossible to access, for example, the events table and the completions table at the same time.

This would be a huge change and I'm not implying we should make changes here to support this, but it might be helpful to think along these lines when refactoring further.

**Comment by tudor-da** [comment]
> The tests timed out on macOS.

Apparently it's just due to this macOS instance being very slow. I see all the tests taking 3-5x more than on the Linux build. I'll just restart it and check the timings.

File: ledger/ledger-api-common/src/main/scala/com/digitalasset/ledger/api/validation/TransactionServiceResponseValidator.scala [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.api.validation
+
+import com.daml.ledger.api.domain.{LedgerId, LedgerOffset}
+import com.daml.lf.data.Ref.LedgerString
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class TransactionServiceResponseValidator(
```
Comment (stefanobaghino-da) [inline ledger/ledger-api-common/src/main/scala/com/digitalasset/ledger/api/validation/TransactionServiceResponseValidator.scala:11]: Was this just moved here or is there some change to be reviewed?

Comment (stefanobaghino-da) [reply ledger/ledger-api-common/src/main/scala/com/digitalasset/ledger/api/validation/TransactionServiceResponseValidator.scala:11]: Wait, I'm not sure I follow. Why do we validate a response?

File: ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/Update.scala [hunk 2]
```diff
@@ -25,6 +25,12 @@ sealed trait Update extends Product with Serializable {
   def recordTime: Timestamp
 }
 
+/** An update describing metadata changes.
+  *
+  * Metadata changes regard parties, packages, ledger configurations and command rejections.
+  */
+sealed trait MetadataUpdate extends Update
```
Comment (stefanobaghino-da) [inline ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/Update.scala:32]: What's the purpose of this change?

**Review by stefanobaghino-da — APPROVED** [review]
This was very difficult to review. I hope I didn't miss anything. I would recommend you break it down into easily reviewable chunks if possible.

File: ledger/ledger-api-common/src/main/scala/com/digitalasset/ledger/api/validation/TransactionServiceResponseValidator.scala (hunk 1 above)
Comment (tudor-da) [reply ledger/ledger-api-common/src/main/scala/com/digitalasset/ledger/api/validation/TransactionServiceResponseValidator.scala:11]: The LedgerDao and Ledger API offer different guarantees now wrt to consistency. In this sense, the API cannot return a transaction that is past the current ledger end, even though the dao can do it.

File: ledger/participant-integration-api/src/main/scala/platform/indexer/ExecuteUpdate.scala [hunk 3]
```diff
@@ -0,0 +1,416 @@
...
+      Timed.future(
+        metrics.daml.index.db.storeTransactionEvents,
+        ledgerDao
+          .storeTransactionEvents(preparedInsert)
+          .map(_ => offsetUpdate),
+      )
+    case offsetUpdate => Future.successful(offsetUpdate)
+  }
+
+  private[indexer] val flow: ExecuteUpdateFlow =
```
[the remark this answers is not in the document]
Comment (rautenrieth-da) [reply ledger/participant-integration-api/src/main/scala/platform/indexer/ExecuteUpdate.scala:297]: It might be worth adding something like `com.daml.metrics.Timed.timedAndTrackedFuture`, but for Akka streams. 

Another use case is measuring the number of elements being currently processed, or measuring the number of items in a `.buffer()` (cc @nmarton-da).

This is not something for this PR though.

File: ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/Update.scala (hunk 2 above)
Comment (tudor-da) [reply ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/Update.scala:32]: I have added it for making threading through the pipeline stages more type safe and modular (e.g. in the pipelined update flow, we either act on the `MetadataUpdate` or on the `TransactionAccepted`).

File: ledger/participant-integration-api/src/main/scala/platform/indexer/ExecuteUpdate.scala [hunk 4]
```diff
@@ -0,0 +1,416 @@
...
+    Flow[OffsetUpdate]
+      .mapAsync(1)(prepareUpdate)
+      .mapAsync(1)(insertTransactionState)
+      .mapAsync(1)(insertTransactionEvents)
```
[the remark this answers is not in the document]
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/indexer/ExecuteUpdate.scala:301]: Indeed, this is a way to go forward.

[the remark this answers is not in the document]
Comment (rautenrieth-da) [reply ledger/participant-integration-api/src/main/scala/platform/indexer/ExecuteUpdate.scala:301]: We could batch through a [.conflate()](https://doc.akka.io/docs/akka/current/stream/operators/Source-or-Flow/conflate.html) stream operator. 

But I think this would be counter-productive. The point of this PR is to allow all three updates to run asynchronously, so that the next transaction can start writing contracts as soon as the first transaction is done with contracts and starts writing events.

File: ledger/ledger-api-common/src/main/scala/com/digitalasset/ledger/api/validation/TransactionServiceResponseValidator.scala (hunk 1 above)
Comment (rautenrieth-da) [reply ledger/ledger-api-common/src/main/scala/com/digitalasset/ledger/api/validation/TransactionServiceResponseValidator.scala:11]: This validator only checks that the returned transaction is not beyond the stored ledger end (which would be an _extremely_ rare case), right?

The corresponding SQL queries already join the events table with the parameters table, in order to check that the transaction was not pruned: 

```
SELECT event_offset, ...
FROM participant_events
JOIN parameters on participant_pruned_up_to_inclusive is null or event_offset > participant_pruned_up_to_inclusive
WHERE transaction_id = $transactionId
```

Could simply modify the SQL query to also check that the event offset is not beyond the stored ledger end?

File: ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/Update.scala (hunk 2 above)
Comment (rautenrieth-da) [reply ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/Update.scala:32]: This seems to split ledger updates to "transaction accepted" and "all other changes". This is related to a particular indexer optimization, and I'm not sure such a trait belongs to the participant state interface.

Before this PR, we had:
```
OffsetUpdate = PreparedTransactionInsert | OffsetStepUpdatePair
```
where `OffsetStepUpdatePair` could include any update (including transactions). It looks like you added `MetadataUpdate` so that you have:
```
OffsetUpdate = (PreparedTransactionInsert | MetadataUpdateStep) | OffsetStepUpdatePair
```
where `MetadataUpdateStep` is the same as `OffsetStepUpdatePair`, except that it's guaranteed to not include transactions. Do we really need that split? If anything, shouldn't it be:
```
OffsetUpdate = PreparedTransactionInsert | UnpreparedUpdateThatIsAnythingButATransactionInsert
```
?

The main benefit seems to be to avoid duplicating the code that handles transaction updates. Before this PR, we had duplicated code and a logger warning for this: https://github.com/digital-asset/daml/pull/8571/files#diff-63690f91ba83940aae6dc820d83d141caec268ff769438437ee49703f3a9d238L391.

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/EventsTablePostgresql.scala [hunk 5]
```diff
@@ -176,9 +176,11 @@ object EventsTablePostgresql extends EventsTable {
     val exerciseChildEventIds = "exerciseChildEventIds"
   }
 
-  private val insertStmt = {
+  private val conflictActionClause = if (idempotentEventInsertions) "on conflict do nothing" else ""
```
Comment (rautenrieth-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/dao/events/EventsTablePostgresql.scala:179]: This ignores all conflicts. Could you add a comment what potential conflicts there are, and why it's ok to ignore all of them? It's hard to look up what constraints and indices we have on the `participant_events` table.

**Review by rautenrieth-da — COMMENTED** [review]
First batch of review comments.

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala [hunk 6]
```diff
@@ -456,37 +488,55 @@ private class JdbcLedgerDao(
...
+      offsetStep: OffsetStep,
+  )(implicit connection: Connection): Unit =
+    Timed.value(
+      metrics.daml.index.db.storeTransactionDbMetrics.insertCompletion,
+      submitterInfo
+        .map(prepareCompletionInsert(_, offsetStep.offset, transactionId, recordTime))
+        .foreach(_.execute()),
+    )
+
+  private def updateLedgerEnd(offsetStep: OffsetStep)(implicit connection: Connection): Unit =
```
Comment (rautenrieth-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala:534]: FYI: At one point, we had the convention that methods that take an `implicit connection: Connection` use the `Sync` suffix to make it easier to see that they don't open an extra database connection. I see only `stopDeduplicatingCommandSync` survived the many refactorings, so there's no point keeping the convention.

File: ledger/participant-integration-api/src/test/lib/scala/platform/store/dao/JdbcIdempotentInsertionsSpec.scala [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+import org.scalatest.flatspec.AsyncFlatSpec
+import org.scalatest.matchers.should.Matchers
+import org.scalatest.{Inside, LoneElement, OptionValues}
+
+trait JdbcIdempotentInsertionsSpec extends Inside with OptionValues with Matchers with LoneElement {
+  self: AsyncFlatSpec with JdbcLedgerDaoSuite =>
+
+  behavior of "JdbcLedgerDao (on PostgreSQL)"
+
+  it should "allow idempotent transaction insertions" in {
```
Comment (rautenrieth-da) [inline ledger/participant-integration-api/src/test/lib/scala/platform/store/dao/JdbcIdempotentInsertionsSpec.scala:19]: Nice to see you are including many tests! If you want to improve it further, you could add the following test(s):

1. Test idempotency of partial inserts. E.g., first insert only the state of a transaction (manually simulating a crash before the events were inserted), then insert only the state + events of the same transaction, then insert the whole transaction.
2. Test that the ledger end guards are working. E.g., first insert only the state + events of a transaction (manually simulating a crash before the ledger end was updated), then verify that this transaction can't be looked up.

File: ledger/ledger-api-common/src/main/scala/com/digitalasset/ledger/api/validation/TransactionServiceResponseValidator.scala (hunk 1 above)
Comment (tudor-da) [reply ledger/ledger-api-common/src/main/scala/com/digitalasset/ledger/api/validation/TransactionServiceResponseValidator.scala]: Done.

File: ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/Update.scala (hunk 2 above)
Comment (tudor-da) [reply ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/Update.scala]: As it is a more widely-used API, I've removed the change.

File: ledger/participant-integration-api/src/main/scala/platform/indexer/ExecuteUpdate.scala [hunk 8]
```diff
@@ -0,0 +1,416 @@
...
+      .mapAsync(1) { case offsetUpdate @ OffsetUpdate(offsetStep, update) =>
+        withEnrichedLoggingContext(loggingContextFor(offsetStep.offset, update)) {
+          implicit loggingContext =>
+            Timed.future(
+              metrics.daml.indexer.stateUpdateProcessing,
+              executeUpdate(offsetUpdate),
+            )
+        }
+      }
```
[the remark this answers is not in the document]
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/indexer/ExecuteUpdate.scala]: Done

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/EventsTablePostgresql.scala (hunk 5 above)
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/events/EventsTablePostgresql.scala]: Done. Added a doc specifying the conflict clause and why is needed.

File: ledger/participant-integration-api/src/test/lib/scala/platform/store/dao/JdbcIdempotentInsertionsSpec.scala (hunk 7 above)
Comment (tudor-da) [reply ledger/participant-integration-api/src/test/lib/scala/platform/store/dao/JdbcIdempotentInsertionsSpec.scala]: Done. Added the tests in `JdbcPipelinedInsertionsSpec`

File: ledger/participant-integration-api/src/main/scala/platform/indexer/ExecuteUpdate.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/indexer/ExecuteUpdate.scala:373]: Added back the timer by threading the timer context through the stages. @rautenrieth-da A reason for which I did not want to time the entire flow is its inherent branching: we would be timing both metadata and pipelined transaction updates on the `store_ledger_entry` timer.

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/EventsTablePostgresql.scala [hunk 9]
```diff
@@ -176,9 +176,19 @@ object EventsTablePostgresql extends EventsTable {
     val exerciseChildEventIds = "exerciseChildEventIds"
   }
 
-  private val insertStmt = {
+  /** Conflict detection relies on the `event_id` primary key only.
```
Comment (rautenrieth-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/dao/events/EventsTablePostgresql.scala:179]: This sounds a bit misleading.

If the indexer does perform post-commit validation (daml-on-sql), then it does so through [ActiveLedgerStateManager](https://github.com/digital-asset/daml/blob/main/ledger/participant-integration-api/src/main/scala/platform/store/ActiveLedgerStateManager.scala) and explicit SQL queries.

If the indexer does not perform post-commit validation, then we expect all writes to be free of conflicts because the underlying ReadService should not emit conflicting transactions.

All indices and constraints in the index database should be either performance optimizations or safeguards that are not strictly necessary.

File: ledger/participant-integration-api/src/test/lib/scala/platform/store/dao/JdbcPipelinedInsertionsSpec.scala [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+  }
+
+  private def assertNoResponseFor(
+      offsetTx: (Offset, LedgerEntry.Transaction)
+  ): Future[Assertion] = {
+    val (offset, tx) = offsetTx
+    val maybeSubmitterInfo = submitterInfo(tx)
+    val preparedInsert = prepareInsert(maybeSubmitterInfo, tx, CurrentOffset(offset))
+    for {
+      _ <- ledgerDao.storeTransactionEvents(preparedInsert)
```
Comment (rautenrieth-da) [inline ledger/participant-integration-api/src/test/lib/scala/platform/store/dao/JdbcPipelinedInsertionsSpec.scala:79]: You could even write one more step:
```suggestion
      _ <- ledgerDao.storeTransactionEvents(preparedInsert)
      _ <- ledgerDao.storeTransactionState(preparedInsert)
```

**Review by rautenrieth-da — APPROVED** [review]
This was a bit difficult to review due to the size. The important parts look good:
- The underlying SQL queries for fetching transactions are guarded against access beyond the ledger end
- There are tests that prove that partially written transactions can be safely re-written
- There are tests that prove that you can't fetch transactions beyond the serialized ledger end (which would be an extreme edge case)

File: ledger/participant-integration-api/src/test/lib/scala/platform/store/dao/JdbcPipelinedInsertionsSpec.scala (hunk 10 above)
Comment (tudor-da) [reply ledger/participant-integration-api/src/test/lib/scala/platform/store/dao/JdbcPipelinedInsertionsSpec.scala:79]: Done

Comment (tudor-da) [reply ledger/participant-integration-api/src/test/lib/scala/platform/store/dao/JdbcPipelinedInsertionsSpec.scala:79]: Done

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/EventsTablePostgresql.scala (hunk 9 above)
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/events/EventsTablePostgresql.scala]: Ok, my aim intent was not clear then. I was referring to table-level insertion conflict (when inserting a row with an already-existing `event_id`) which is not a _update conflict_ for the pipelined indexer and should just be ignored. I've adapted the comment.
